// @compiled/babel-plugin/dist/keyframes/__fixtures__/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.shorthandStyledTaggedTemplateExpression = exports.shorthandStyledObjectCallExpression = exports.shorthandCssPropTaggedTemplateExpression = exports.shorthandCssPropObjectCallExpression = exports.longhandStyledTaggedTemplateExpression = exports.longhandStyledObjectCallExpression = exports.longhandCssPropTaggedTemplateExpression = exports.longhandCssPropObjectCallExpression = void 0;
exports.longhandCssPropObjectCallExpression = `
  <div css={{
    animationDuration: '2s',
    animationName: fadeOut,
    animationTimingFunction: 'ease-in-out',
  }} />
`;
exports.longhandCssPropTaggedTemplateExpression = `
  <div css={css\`
    animation-duration: 2s;
    animation-name: \${fadeOut};
    animation-timing-function: ease-in-out;
  \`} />
`;
exports.longhandStyledObjectCallExpression = `
  const StyledComponent = styled.div({
    animationDuration: '2s',
    animationName: fadeOut,
    animationTimingFunction: 'ease-in-out',
  });
`;
exports.longhandStyledTaggedTemplateExpression = `
  const StyledComponent = styled.div\`
    animation-duration: 2s;
    animation-name: \${fadeOut};
    animation-timing-function: ease-in-out;
  \`;
`;
exports.shorthandCssPropObjectCallExpression = `<div css={{ animation: \`\${fadeOut} 2s ease-in-out\` }} />`;
exports.shorthandCssPropTaggedTemplateExpression = `<div css={css\`animation: \${fadeOut} 2s ease-in-out\`} />`;
exports.shorthandStyledObjectCallExpression = `
  const StyledComponent = styled.div({
    animation: \`\${fadeOut} 2s ease-in-out\`,
  });
`;
exports.shorthandStyledTaggedTemplateExpression = `
  const StyledComponent = styled.div\`
    animation: \${fadeOut} 2s ease-in-out;
  \`;
`;
