// @atlaskit/eslint-plugin-design-system/dist/esm/rules/no-physical-properties/index.js
import { isNodeOfType } from 'eslint-codemod-utils';
import { createLintRule } from '../utils/create-rule';
import { physicalLogicalMap } from './logical-physical-map';
var rule = createLintRule({
  meta: {
    name: 'no-physical-properties',
    fixable: 'code',
    docs: {
      description: 'Disallow physical properties and values in `css` function calls.',
      recommended: false,
      severity: 'error'
    },
    messages: {
      noPhysicalProperties: 'Physical properties are not allowed in `css` functions as they do not support different reading modes. Use a logical property instead.',
      noPhysicalValues: 'Physical values are not allowed in `css` functions.'
    }
  },
  create: function create(context) {
    return {
      'CallExpression[callee.name=css] > ObjectExpression Property,CallExpression[callee.name=xcss] > ObjectExpression Property': function CallExpressionCalleeNameCss__ObjectExpression_PropertyCallExpressionCalleeNameXcss__ObjectExpression_Property(node) {
        if (!isNodeOfType(node, 'Property')) {
          return;
        }
        if (!isNodeOfType(node.key, 'Identifier')) {
          return;
        }
        var key = node.key;
        if (key.name in physicalLogicalMap) {
          context.report({
            node: key,
            messageId: 'noPhysicalProperties',
            fix: function fix(fixer) {
              var logicalProperty = physicalLogicalMap[key.name];
              return fixer.replaceText(key, logicalProperty);
            }
          });
        }
      }
    };
  }
});
export default rule;
