// @atlaskit/navigation-system/dist/esm/ui/top-nav-items/nav-logo/logo-renderer.js
/* logo-renderer.tsx generated by @compiled/babel-plugin v0.36.1 */
import { ax, ix } from "@compiled/react/runtime";
import React, { memo } from 'react';
import { fg } from '@atlaskit/platform-feature-flags';
import { useHasCustomTheme } from '../themed/has-custom-theme-context';
export var themedLogoIcon = '--ds-top-bar-logo-icon';
export var themedLogoText = '--ds-top-bar-logo-text';
export function LogoRenderer(_ref) {
  var logoOrIcon = _ref.logoOrIcon;
  return fg('jiv-20710-fix-nav-rerender') ? /*#__PURE__*/React.createElement(LogoRendererMemo, {
    logoOrIcon: logoOrIcon
  }) : /*#__PURE__*/React.createElement(LogoRendererNoMemo, {
    logoOrIcon: logoOrIcon
  });
}
var LogoRendererMemo = /*#__PURE__*/memo(LogoRendererNoMemo);
function LogoRendererNoMemo(_ref2) {
  var LogoOrIcon = _ref2.logoOrIcon;
  var hasCustomTheme = useHasCustomTheme();
  if (hasCustomTheme) {
    return /*#__PURE__*/React.createElement(LogoOrIcon, {
      size: "small",
      label: "",
      iconColor: "var(".concat(themedLogoIcon, ")"),
      textColor: "var(".concat(themedLogoText, ")")
    });
  }
  return /*#__PURE__*/React.createElement(LogoOrIcon, {
    size: "small",
    label: "",
    appearance: "brand"
  });
}
